File: .agent/skills/cnma-cap-backend/templates/java/srv/src/main/java/cnma/{{module_name}}/infrastructure/database/DBHandler.java
package cnma.{{module_name}}.infrastructure.database;

import com.sap.cds.Result;
import com.sap.cds.services.persistence.PersistenceService;
import com.sap.cds.ql.Select;
import com.sap.cds.ql.Insert;
import com.sap.cds.ql.Update;
import com.sap.cds.ql.Delete;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * DBHandler - Generic database operations with transaction management.
 * DIP: Domain layer depends on this abstraction, not concrete CQN.
 */
@Slf4j
@Component
public class DBHandler {

    private final PersistenceService persistence;

    public DBHandler(PersistenceService persistence) {
        this.persistence = persistence;
    }

    /**
     * Execute SELECT query with conditions.
     */
    public <T> List<T> select(Class<T> entityClass, String entityName, Map<String, Object> conditions) {
        try {
            Select<?> builder = Select.from(entityName);
            if (conditions != null && !conditions.isEmpty()) {
                builder.where(conditions);
            }
            Result result = persistence.run(builder);
            return result.stream(entityClass).collect(Collectors.toList());
        } catch (Exception e) {
            log.error("SELECT failed for {}", entityName, e);
            return List.of();
        }
    }

    /**
     * Find by ID.
     */
    public <T> Optional<T> findById(Class<T> entityClass, String entityName, String id) {
        try {
            Result result = persistence.run(
                Select.from(entityName).where("ID", id)
            );
            return result.first(entityClass);
        } catch (Exception e) {
            log.error("findById failed for {} with id {}", entityName, id, e);
            return Optional.empty();
        }
    }

    /**
     * Insert single record.
     */
    public <T> T insert(String entityName, T data) {
        try {
            persistence.run(Insert.into(entityName).entry(data));
            return data;
        } catch (Exception e) {
            log.error("INSERT failed for {}", entityName, e);
            throw new RuntimeException("Insert failed: " + e.getMessage(), e);
        }
    }

    /**
     * Insert multiple records (batch).
     */
    public <T> void insertBatch(String entityName, List<T> entries) {
        try {
            persistence.run(Insert.into(entityName).entries(entries));
        } catch (Exception e) {
            log.error("Batch INSERT failed for {}", entityName, e);
            throw new RuntimeException("Batch insert failed: " + e.getMessage(), e);
        }
    }

    /**
     * Update records matching conditions.
     */
    public int update(String entityName, Map<String, Object> conditions, Map<String, Object> data) {
        try {
            Update<?> builder = Update.entity(entityName).data(data);
            if (conditions != null && !conditions.isEmpty()) {
                builder.where(conditions);
            }
            Result result = persistence.run(builder);
            return result.rowCount();
        } catch (Exception e) {
            log.error("UPDATE failed for {}", entityName, e);
            throw new RuntimeException("Update failed: " + e.getMessage(), e);
        }
    }

    /**
     * Delete records matching conditions.
     */
    public void delete(String entityName, Map<String, Object> conditions) {
        try {
            Delete<?> builder = Delete.from(entityName);
            if (conditions != null && !conditions.isEmpty()) {
                builder.where(conditions);
            }
            persistence.run(builder);
        } catch (Exception e) {
            log.error("DELETE failed for {}", entityName, e);
            throw new RuntimeException("Delete failed: " + e.getMessage(), e);
        }
    }

    /**
     * Execute custom CQN query.
     */
    public Result execute(Select<?> query) {
        return persistence.run(query);
    }
}

File: .agent/skills/cnma-cap-backend/templates/node/srv/src/infrastructure/database/DBHandler.ts
import cds from '@sap/cds';
import { SELECT, INSERT, UPDATE, DELETE } from '@sap/cds/pg';
import { createLogger } from '../../common/util/Logger';
import { HttpStatusCode } from '../../common/enum/HttpStatusCodeEnum';
import { ServiceResponse } from '../../common/model/core/ApiResponse';

const logger = createLogger('DBHandler');

/**
 * DBHandler - Generic database operations with transaction management.
 *
 * This belongs in Infrastructure layer because:
 * - Manages database connections and transactions
 * - Handles low-level CQN operations
 * - Domain layer should not know about DB implementation
 *
 * Usage: Repository classes extend this to inherit common CRUD with transactions.
 */
export abstract class DBHandler {
    protected tx: cds.Transaction;

    constructor() {
        this.tx = cds.tx(cds.context);
    }

    /**
     * Execute SELECT query with optional conditions.
     */
    protected async select<T>(
        entity: any,
        conditions?: any,
        options?: { columns?: any[]; limit?: number; skip?: number }
    ): Promise<T[]> {
        let query = SELECT.from(entity);

        if (conditions && conditions !== '*') {
            query = query.where(conditions);
        }

        if (options?.columns) {
            query = query.columns(options.columns);
        }

        if (options?.limit !== undefined) {
            query = query.limit(options.limit, options.skip ?? 0);
        }

        const results = await this.tx.run(query);
        await this.tx.commit();
        return results as T[];
    }

    /**
     * Insert single record.
     */
    protected async insert<T>(entity: any, data: any): Promise<T> {
        const results = await this.tx.run(INSERT.into(entity).entries(data));
        await this.tx.commit();
        return results as T;
    }

    /**
     * Insert multiple records (batch).
     */
    protected async insertBatch<T>(entity: any, entries: any[]): Promise<T[]> {
        const results = await this.tx.run(INSERT.into(entity).entries(entries));
        await this.tx.commit();
        return results as T[];
    }

    /**
     * Update records matching conditions.
     */
    protected async update(entity: any, conditions: any, data: any): Promise<number> {
        const result = await this.tx.run(
            UPDATE.entity(entity).where(conditions).data(data)
        );
        await this.tx.commit();
        return result;
    }

    /**
     * Delete records matching conditions.
     */
    protected async delete(entity: any, conditions: any): Promise<void> {
        await this.tx.run(DELETE.from(entity).where(conditions));
        await this.tx.commit();
    }

    /**
     * Execute custom CQN query.
     */
    protected async execute(query: any): Promise<any> {
        const results = await this.tx.run(query);
        await this.tx.commit();
        return results;
    }

    /**
     * Execute with manual transaction control.
     */
    protected async executeWithTx<T>(fn: () => Promise<T>): Promise<T> {
        try {
            const result = await fn();
            await this.tx.commit();
            return result;
        } catch (error) {
            await this.tx.rollback().catch((err) => {
                logger.error('Rollback error', { error: err });
            });
            throw error;
        }
    }

    /**
     * Rollback on error (call in catch block).
     */
    protected async rollback(error?: any): Promise<void> {
        await this.tx.rollback().catch((err) => {
            logger.error('Rollback error', { error: err });
        });
        if (error) {
            logger.error('Transaction failed', { error });
        }
    }

    /**
     * Build ServiceResponse for get operations.
     */
    protected buildGetResponse<T>(results: T[]): ServiceResponse<T> {
        const statusCode = results && results.length > 0
            ? HttpStatusCode.OK
            : HttpStatusCode.NOT_FOUND;
        const message = results && results.length > 0
            ? 'Data retrieved successfully'
            : 'No data found';
        return new ServiceResponse(statusCode, message, results);
    }
}